src/Microsoft.ML.Data/Transforms/Hashing.cs

```diff
@@ -834,6 +1112,96 @@ private void AddMetaKeyValues(int i, DataViewSchema.Annotations.Builder builder)
+                    castNode.AddAttribute("to", NumberDataViewType.UInt32.RawType);
+                    murmurNode = ctx.CreateNode(opType, castOutput, murmurOutput, ctx.GetNodeName(opType), "com.microsoft");
+                }
+                else if (srcType == typeof(uint) || srcType == typeof(int) || srcType == typeof(ReadOnlyMemory<char>))
+                {
+                    murmurNode = ctx.CreateNode(opType, srcVariable, murmurOutput, ctx.GetNodeName(opType), "com.microsoft");
+                }
+                else
+                {
+                    return false;
+                }
+
+                murmurNode.AddAttribute("positive", 1);
+                var seed = _parent._columns[iinfo].Seed;
+                if (_parent._columns[iinfo].UseOrderedHashing)
+                    seed = Hashing.MurmurRound(seed, 0);
+                murmurNode.AddAttribute("seed", seed);
+
+                // masking is done via bitshifts, until bitwise AND is supported by Onnxruntime
+                opType = "BitShift";
+                string bitShiftOutput = ctx.AddIntermediateVariable(_dstTypes[iinfo], "BitShiftOutput");
+                var shiftValue = ctx.AddInitializer((ulong)(32 - _parent._columns[iinfo].NumberOfBits), false);
+                var shiftNode = ctx.CreateNode(opType, new[] { murmurOutput, shiftValue }, new[] { bitShiftOutput }, ctx.GetNodeName(opType), "");
+                shiftNode.AddAttribute("direction", "LEFT");
+
+                opType = "BitShift";
+                var outputVariable = dstVariable;
+                if(_srcTypes[iinfo].GetItemType() is KeyDataViewType)
+                    outputVariable = ctx.AddIntermediateVariable(_dstTypes[iinfo], "BitShiftOutput2");
+                shiftNode = ctx.CreateNode(opType, new[] { bitShiftOutput, shiftValue }, new[] { outputVariable }, ctx.GetNodeName(opType), "");
+                shiftNode.AddAttribute("direction", "RIGHT");
+
+                if (_srcTypes[iinfo].GetItemType() is KeyDataViewType)
+                {
+                    opType = "Cast";
+                    string castOutput = ctx.AddIntermediateVariable(NumberDataViewType.Int64, "CastOutput2");
+                    var castNode = ctx.CreateNode(opType, outputVariable, castOutput, ctx.GetNodeName(opType), "");
+                    var t = NumberDataViewType.Int64.RawType;
+                    castNode.AddAttribute("to", t);
+
+                    opType = "Add";
+                    string addOutput = ctx.AddIntermediateVariable(NumberDataViewType.Int64, "AddOutput");
+                    string one = ctx.AddInitializer(1);
+                    ctx.CreateNode(opType, new[] { castOutput, one }, new[] { addOutput }, ctx.GetNodeName(opType), "");
+
+                    opType = "Cast";
+                    var castNodeFinal = ctx.CreateNode(opType, addOutput, dstVariable, ctx.GetNodeName(opType), "");
+                    var tFinal = NumberDataViewType.UInt32.RawType;
```

Comment:
>tFinal [](start = 24, length = 6)

If I understand correctly, these three nodes are supposed to add 1 to the output, and then convert back to the original type. But key types are not always uint, they can also be other unsigned integer types.

Reply:
updated!

Reply:
So after looking at it again, my previous comment was inaccurate. The `HashingTransformer` **always** outputs a key type, so these additional nodes that you are adding (Cast=>Add 1=>Cast back) actually have to be added in **all** cases, not only when the input is a key type. On the other hand, the raw type of the output of `HashingTransformer` is always `uint` (see the [`GetOutputType`](https://github.com/dotnet/machinelearning/blob/master/src/Microsoft.ML.Data/Transforms/Hashing.cs#L149) ), so the previous way you did the Cast operator was correct.

The difference between a key type input and a non-key unsigned integer type input, is that for keys a value of 0 is mapped to 0 (0 is the `NA` value of key types, and for regular unsigned integers it is hashed just like any other value. For example, the uint hashing function is

```
            public uint HashCore(uint seed, uint mask, in uint value)
                => (Hashing.MixHash(Hashing.MurmurRound(seed, value)) & mask) + 1;
```

and the uint key type hashing function is 

```
            public uint HashCore(uint seed, uint mask, in uint value)
                => value == 0 ? 0 : (Hashing.MixHash(Hashing.MurmurRound(seed, value)) & mask) + 1;
```

Mapping 0 to 0 is currently not handled at all in the ONNX export, not sure if it can be handled... we might need to return `false` for key type inputs if this cannot be handled, similarly to what you do for floats and doubles.


---
In reply to: [414083637](https://github.com/dotnet/machinelearning/pull/5013#discussion_r414083637) [](ancestors = 414083637)